File: rider-extension/src/main/kotlin/com/yeahbah/efvibe/services/CliRunner.kt
package com.yeahbah.efvibe.services

import com.intellij.execution.configurations.GeneralCommandLine
import com.intellij.execution.process.CapturingProcessHandler
import com.intellij.execution.process.ProcessOutput
import com.intellij.openapi.components.service
import com.intellij.openapi.project.Project
import java.nio.charset.StandardCharsets
import kotlin.io.path.exists

class CliRunner(private val project: Project) {
    private val settings: EfvibeSettingsState get() = project.service<EfvibeSettingsService>().state

    fun buildReplCommandLine(): String {
        val invocation = resolveInvocation()
        return listOf(invocation.command)
            .plus(invocation.prefixArgs)
            .plus(baseArgs())
            .joinToString(" ") { quote(it) }
    }

    fun runAboutJson(): CliResult =
        run(listOf("--about-json", "--no-banner"), timeoutMs = 30_000)

    fun runExpression(expression: String, withPlan: Boolean): CliResult {
        val args = baseArgs().toMutableList()
        args += listOf("-e", expression, "--format", "json", "--no-banner")
        if (withPlan) args += "--with-plan"
        return run(args, timeoutMs = 10 * 60_000)
    }

    fun runExpressionPayload(expression: String, withPlan: Boolean, preferDaemon: Boolean = true): ExpressionRunResult {
        if (preferDaemon) {
            val daemonFailure = runCatching {
                return EfvibeDaemonClient.getInstance(project).runExpression(expression, withPlan)
            }.exceptionOrNull()

            if (daemonFailure != null) {
                val fallback = runExpression(expression, withPlan)
                return ExpressionRunResult(
                    result = fallback,
                    payload = EfvibeJsonParser.parseEvaluation(fallback.stdout),
                    usedDaemon = false,
                    daemonError = daemonFailure.message ?: daemonFailure.toString(),
                )
            }
        }

        val result = runExpression(expression, withPlan)
        return ExpressionRunResult(result, EfvibeJsonParser.parseEvaluation(result.stdout))
    }

    fun runDbInfo(): CliResult =
        run(baseArgs() + listOf("--dbinfo-json", "--no-banner"), timeoutMs = 10 * 60_000)

    fun runDbInfoPayload(preferDaemon: Boolean = true): DaemonJsonResult<DbInfoPayload?> =
        runJsonWithDaemon(
            preferDaemon = preferDaemon,
            daemonRequest = { EfvibeDaemonClient.getInstance(project).runDbInfo() },
            fallback = { runDbInfo() },
            parse = { EfvibeJsonParser.parseDbInfo(it) },
        )

    fun runTables(): CliResult =
        run(baseArgs() + listOf("--tables-json", "--no-banner"), timeoutMs = 10 * 60_000)

    fun runTablesPayload(preferDaemon: Boolean = true): DaemonJsonResult<TablesPayload?> =
        runJsonWithDaemon(
            preferDaemon = preferDaemon,
            daemonRequest = { EfvibeDaemonClient.getInstance(project).runTables() },
            fallback = { runTables() },
            parse = { EfvibeJsonParser.parseTables(it) },
        )

    fun runDescribe(entityName: String, preferDaemon: Boolean = true): DaemonRunResult =
        runWithDaemon(
            preferDaemon = preferDaemon,
            daemonRequest = { EfvibeDaemonClient.getInstance(project).runDescribe(entityName) },
            fallback = { runDescribeCli(entityName) },
        )

    private fun runDescribeCli(entityName: String): CliResult =
        run(baseArgs() + listOf("--describe-json", entityName, "--no-banner"), timeoutMs = 10 * 60_000)

    fun runScan(mode: String): CliResult {
        val args = mutableListOf("scan", mode)
        args += baseArgs()
        args += listOf("--json", "--no-banner")
        if (settings.scanRespectDismissals) args += "--respect-dismissals"
        if (settings.scanMinSeverity.isNotBlank()) args += listOf("--min-severity", settings.scanMinSeverity.trim())
        return run(args, timeoutMs = 20 * 60_000)
    }

    fun runScanPayload(mode: String, preferDaemon: Boolean = true): DaemonJsonResult<ScanPayload?> =
        runJsonWithDaemon(
            preferDaemon = preferDaemon,
            daemonRequest = {
                EfvibeDaemonClient.getInstance(project).runScan(
                    mode,
                    settings.scanRespectDismissals,
                    settings.scanMinSeverity,
                )
            },
            fallback = { runScan(mode) },
            parse = { EfvibeJsonParser.parseScan(it) },
        )

    fun runScanNote(finding: ScanFinding, note: String): CliResult =
        run(scanFindingArgs("note", finding) + listOf("--text", note), timeoutMs = 30_000)

    fun runScanDismiss(finding: ScanFinding, note: String?): CliResult {
        val args = scanFindingArgs("dismiss", finding).toMutableList()
        if (!note.isNullOrBlank()) {
            args += listOf("--note", note.trim())
        }
        return run(args, timeoutMs = 30_000)
    }

    fun buildServeSpec(): CliCommandSpec =
        buildSpec(listOf("serve") + baseArgs())

    fun buildSpec(args: List<String>): CliCommandSpec {
        val invocation = resolveInvocation()
        return CliCommandSpec(
            command = invocation.command,
            args = invocation.prefixArgs + args,
            workingDirectory = PathResolver.solutionDirectory(project).toFile(),
        )
    }

    private inline fun <T> runJsonWithDaemon(
        preferDaemon: Boolean,
        crossinline daemonRequest: () -> String,
        fallback: () -> CliResult,
        crossinline parse: (String) -> T?,
    ): DaemonJsonResult<T> {
        if (preferDaemon) {
            val failure = runCatching {
                val stdout = daemonRequest()
                return DaemonJsonResult(
                    result = CliResult(0, stdout, ""),
                    payload = parse(stdout),
                    usedDaemon = true,
                )
            }.exceptionOrNull()

            if (failure != null) {
                val cli = fallback()
                return DaemonJsonResult(
                    result = cli,
                    payload = parse(cli.stdout),
                    usedDaemon = false,
                    daemonError = failure.message ?: failure.toString(),
                )
            }
        }

        val cli = fallback()
        return DaemonJsonResult(cli, parse(cli.stdout))
    }

    private inline fun runWithDaemon(
        preferDaemon: Boolean,
        crossinline daemonRequest: () -> String,
        fallback: () -> CliResult,
    ): DaemonRunResult {
        if (preferDaemon) {
            val failure = runCatching {
                val stdout = daemonRequest()
                return DaemonRunResult(CliResult(0, stdout, ""), usedDaemon = true)
            }.exceptionOrNull()

            if (failure != null) {
                val cli = fallback()
                return DaemonRunResult(
                    result = cli,
                    usedDaemon = false,
                    daemonError = failure.message ?: failure.toString(),
                )
            }
        }

        return DaemonRunResult(fallback())
    }

    private fun run(args: List<String>, timeoutMs: Int): CliResult {
        val spec = buildSpec(args)
        val commandLine = GeneralCommandLine(spec.command)
            .withParameters(spec.args)
            .withWorkDirectory(spec.workingDirectory)
            .withCharset(StandardCharsets.UTF_8)

        val output: ProcessOutput = CapturingProcessHandler(commandLine).runProcess(timeoutMs)

        return CliResult(output.exitCode, output.stdout, output.stderr)
    }

    private fun resolveInvocation(): CliInvocation {
        val toolPath = PathResolver.resolve(settings.toolPath, project)
        if (toolPath.isNotBlank() && java.nio.file.Path.of(toolPath).exists()) {
            return CliInvocation(toolPath)
        }

        if (findDotnetToolsManifest() != null) {
            val framework = settings.dotnetFramework.trim()
            val prefixArgs = if (framework.isBlank()) listOf("efvibe") else listOf("efvibe", "-f", framework)
            return CliInvocation("dotnet", prefixArgs)
        }

        return CliInvocation("efvibe")
    }

    private fun baseArgs(): List<String> = buildList {
        addOption("-w", PathResolver.resolve(settings.workspaceRoot, project))
        addOption("-p", PathResolver.resolve(settings.project, project))
        addOption("-s", PathResolver.resolve(settings.startupProject, project))
        addOption("-c", settings.context.trim())
        addOption("--connection-string", settings.connectionString.trim())
        addOption("--provider", settings.provider.trim())
        if (!settings.dbLog) add("--no-dblog")
        addOption("--framework", settings.dotnetFramework.trim())
    }

    private fun MutableList<String>.addOption(name: String, value: String) {
        if (value.isBlank()) return
        add(name)
        add(value)
    }

    private fun scanFindingArgs(command: String, finding: ScanFinding): List<String> = buildList {
        add("scan")
        add(command)
        addOption("-w", PathResolver.resolve(settings.workspaceRoot, project))
        addOption("-p", PathResolver.resolve(settings.project, project))
        addOption("-c", settings.context.trim())
        addOption("--file", finding.filePath)
        addOption("--line", finding.line.toString())
        addOption("--rule", finding.ruleId)
        addOption("--code", finding.code)
    }

    private fun findDotnetToolsManifest(): java.nio.file.Path? {
        var current: java.nio.file.Path? = PathResolver.solutionDirectory(project)
        repeat(12) {
            val candidates = listOfNotNull(
                current?.resolve("dotnet-tools.json"),
                current?.resolve(".config")?.resolve("dotnet-tools.json"),
            )
            candidates.firstOrNull { it.exists() }?.let { return it }
            current = current?.parent
        }
        return null
    }

    private fun quote(value: String): String =
        if (value.any { it.isWhitespace() || it == '"' }) "\"${value.replace("\"", "\\\"")}\"" else value
}

File: visualstudio-extension/Services/CliRunner.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using MyEfVibe.VisualStudio.Models;

namespace MyEfVibe.VisualStudio.Services;

internal sealed class CliRunner
{
    private readonly string _solutionDirectory;

    public CliRunner(string solutionDirectory)
    {
        _solutionDirectory = solutionDirectory;
    }

    public CliInvocation ResolveInvocation(EfvibeSettings settings)
    {
        if (!string.IsNullOrWhiteSpace(settings.ToolPath))
        {
            return new CliInvocation { Command = settings.ToolPath };
        }

        if (FindDotnetToolsManifest(_solutionDirectory) is not null)
        {
            var invocation = new CliInvocation { Command = "dotnet" };
            invocation.PrefixArgs.Add("efvibe");
            return invocation;
        }

        return new CliInvocation { Command = "efvibe" };
    }

    public IReadOnlyList<string> BuildBaseArgs(EfvibeSettings settings)
    {
        var args = new List<string>();

        AddOption(args, "-w", settings.WorkspaceRoot);
        AddOption(args, "-p", settings.Project);
        AddOption(args, "-s", settings.StartupProject);
        AddOption(args, "-c", settings.Context);
        AddOption(args, "--connection-string", settings.ConnectionString);
        AddOption(args, "--provider", settings.Provider);

        if (!settings.DbLog)
            args.Add("--no-dblog");

        AddOption(args, "--framework", settings.DotnetFramework);

        return args;
    }

    public string BuildReplCommandLine(EfvibeSettings settings)
    {
        var invocation = ResolveInvocation(settings);
        var args = new List<string>(invocation.PrefixArgs);
        args.AddRange(BuildBaseArgs(settings));

        return BuildCommandLine(invocation.Command, args);
    }

    public CliInvocationSpec BuildServeSpec(EfvibeSettings settings)
    {
        var invocation = ResolveInvocation(settings);
        var args = new List<string>(invocation.PrefixArgs);
        args.Add("serve");
        args.AddRange(BuildBaseArgs(settings));

        return new CliInvocationSpec
        {
            Command = invocation.Command,
            Args = args.ToArray(),
            WorkingDirectory = _solutionDirectory,
        };
    }

    public async Task<ExpressionRunResult> RunExpressionPayloadAsync(
        EfvibeWorkspace.WorkspaceContext workspace,
        string expression,
        bool withPlan,
        bool preferDaemon,
        CancellationToken cancellationToken)
    {
        var settings = workspace.Settings;

        if (preferDaemon)
        {
            try
            {
                var daemon = EfvibeDaemonClient.GetOrCreate(workspace);
                return await Task.Run(() => daemon.RunExpression(expression, withPlan), cancellationToken);
            }
            catch (Exception ex)
            {
                var fallback = await RunExpressionJsonAsync(settings, expression, withPlan, cancellationToken);
                return new ExpressionRunResult(
                    fallback,
                    JsonLineParser.ParseFirstJsonLine<EvaluationJsonPayload>(fallback.Stdout),
                    usedDaemon: false,
                    daemonError: ex.Message);
            }
        }

        var result = await RunExpressionJsonAsync(settings, expression, withPlan, cancellationToken);
        return new ExpressionRunResult(
            result,
            JsonLineParser.ParseFirstJsonLine<EvaluationJsonPayload>(result.Stdout));
    }

    public async Task<CliRunResult> RunExpressionJsonAsync(
        EfvibeSettings settings,
        string expression,
        bool withPlan,
        CancellationToken cancellationToken)
    {
        var args = new List<string>(BuildBaseArgs(settings))
        {
            "-e",
            expression,
            "--format",
            "json",
            "--no-banner",
        };

        if (withPlan)
            args.Add("--with-plan");

        return await RunAsync(settings, args, TimeSpan.FromMinutes(10), cancellationToken);
    }

    public Task<CliRunResult> RunDbInfoJsonAsync(EfvibeSettings settings, CancellationToken cancellationToken) =>
        RunJsonFlagAsync(settings, "--dbinfo-json", cancellationToken);

    public Task<CliRunResult> RunTablesJsonAsync(EfvibeSettings settings, CancellationToken cancellationToken) =>
        RunJsonFlagAsync(settings, "--tables-json", cancellationToken);

    public Task<CliRunResult> RunAboutJsonAsync(EfvibeSettings settings, CancellationToken cancellationToken)
    {
        var args = new List<string>
        {
            "--about-json",
            "--no-banner",
        };

        return RunAsync(settings, args, TimeSpan.FromSeconds(30), cancellationToken);
    }

    public Task<CliRunResult> RunDescribeJsonAsync(
        EfvibeSettings settings,
        string entityName,
        CancellationToken cancellationToken)
    {
        var args = new List<string>(BuildBaseArgs(settings))
        {
            "--describe-json",
            entityName,
            "--no-banner",
        };

        return RunAsync(settings, args, TimeSpan.FromMinutes(10), cancellationToken);
    }

    public Task<CliRunResult> RunScanAsync(
        EfvibeSettings settings,
        string mode,
        CancellationToken cancellationToken)
    {
        var args = new List<string>
        {
            "scan",
            mode,
        };

        args.AddRange(BuildBaseArgs(settings));
        args.Add("--json");
        args.Add("--no-banner");

        if (settings.ScanRespectDismissals)
            args.Add("--respect-dismissals");

        AddOption(args, "--min-severity", settings.ScanMinSeverity);

        return RunAsync(settings, args, TimeSpan.FromMinutes(20), cancellationToken);
    }

    public Task<CliRunResult> RunScanNoteAsync(
        EfvibeSettings settings,
        ScanFinding finding,
        string note,
        CancellationToken cancellationToken)
    {
        var args = BuildScanFindingArgs(settings, "note", finding);
        args.Add("--text");
        args.Add(note);
        return RunAsync(settings, args, TimeSpan.FromSeconds(30), cancellationToken);
    }

    public Task<CliRunResult> RunScanDismissAsync(
        EfvibeSettings settings,
        ScanFinding finding,
        string? note,
        CancellationToken cancellationToken)
    {
        var args = BuildScanFindingArgs(settings, "dismiss", finding);
        if (!string.IsNullOrWhiteSpace(note))
        {
            args.Add("--note");
            args.Add(note.Trim());
        }

        return RunAsync(settings, args, TimeSpan.FromSeconds(30), cancellationToken);
    }

    public void StartReplInExternalTerminal(EfvibeSettings settings)
    {
        var commandLine = BuildReplCommandLine(settings);
        var startInfo = new ProcessStartInfo
        {
            FileName = "cmd.exe",
            Arguments = "/k " + QuoteForCmd(commandLine),
            WorkingDirectory = _solutionDirectory,
            UseShellExecute = true,
        };

        Process.Start(startInfo);
    }

    private Task<CliRunResult> RunJsonFlagAsync(
        EfvibeSettings settings,
        string flag,
        CancellationToken cancellationToken)
    {
        var args = new List<string>(BuildBaseArgs(settings))
        {
            flag,
            "--no-banner",
        };

        return RunAsync(settings, args, TimeSpan.FromMinutes(10), cancellationToken);
    }

    private async Task<CliRunResult> RunAsync(
        EfvibeSettings settings,
        IReadOnlyList<string> args,
        TimeSpan timeout,
        CancellationToken cancellationToken)
    {
        var invocation = ResolveInvocation(settings);
        var allArgs = new List<string>(invocation.PrefixArgs);
        allArgs.AddRange(args);

        using var process = new Process();
        process.StartInfo = new ProcessStartInfo
        {
            FileName = invocation.Command,
            Arguments = BuildArguments(allArgs),
            WorkingDirectory = _solutionDirectory,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true,
            StandardOutputEncoding = Encoding.UTF8,
            StandardErrorEncoding = Encoding.UTF8,
        };

        var stdout = new StringBuilder();
        var stderr = new StringBuilder();

        process.OutputDataReceived += (_, e) =>
        {
            if (e.Data is not null)
                stdout.AppendLine(e.Data);
        };
        process.ErrorDataReceived += (_, e) =>
        {
            if (e.Data is not null)
                stderr.AppendLine(e.Data);
        };

        process.Start();
        process.BeginOutputReadLine();
        process.BeginErrorReadLine();

        using var timeoutSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        timeoutSource.CancelAfter(timeout);

        await WaitForExitAsync(process, timeoutSource.Token);

        return new CliRunResult
        {
            ExitCode = process.ExitCode,
            Stdout = stdout.ToString(),
            Stderr = stderr.ToString(),
        };
    }

    private static Task WaitForExitAsync(Process process, CancellationToken cancellationToken)
    {
        if (process.HasExited)
            return Task.CompletedTask;

        var completion = new TaskCompletionSource<object?>();
        process.EnableRaisingEvents = true;
        process.Exited += (_, _) => completion.TrySetResult(null);

        cancellationToken.Register(() =>
        {
            try
            {
                if (!process.HasExited)
                    process.Kill();
            }
            catch (InvalidOperationException)
            {
            }

            completion.TrySetCanceled(cancellationToken);
        });

        return completion.Task;
    }

    private static void AddOption(ICollection<string> args, string name, string value)
    {
        if (string.IsNullOrWhiteSpace(value))
            return;

        args.Add(name);
        args.Add(value.Trim());
    }

    private static string? FindDotnetToolsManifest(string startDirectory)
    {
        var current = new DirectoryInfo(startDirectory);

        for (var depth = 0; current is not null && depth < 12; depth++)
        {
            var candidate = Path.Combine(current.FullName, ".config", "dotnet-tools.json");

            if (File.Exists(candidate))
                return candidate;

            current = current.Parent;
        }

        return null;
    }

    private static string BuildCommandLine(string command, IReadOnlyList<string> args) =>
        QuoteArg(command) + (args.Count == 0 ? string.Empty : " " + BuildArguments(args));

    internal static string BuildArguments(IEnumerable<string> args) =>
        string.Join(" ", args.Select(QuoteArg));

    private List<string> BuildScanFindingArgs(EfvibeSettings settings, string command, ScanFinding finding)
    {
        var args = new List<string> { "scan", command };
        AddOption(args, "-w", settings.WorkspaceRoot);
        AddOption(args, "-p", settings.Project);
        AddOption(args, "-c", settings.Context);
        AddOption(args, "--file", finding.FilePath);
        AddOption(args, "--line", finding.Line.ToString());
        AddOption(args, "--rule", finding.RuleId);
        AddOption(args, "--code", finding.Code);
        return args;
    }

    private static string QuoteArg(string value)
    {
        if (string.IsNullOrEmpty(value))
            return "\"\"";

        return value.IndexOfAny(new[] { ' ', '\t', '"' }) < 0
            ? value
            : "\"" + value.Replace("\"", "\\\"") + "\"";
    }

    private static string QuoteForCmd(string commandLine) =>
        "\"" + commandLine.Replace("\"", "\\\"") + "\"";
}
